NIX_PATH: don't prepend $HOME-based value in session variable, set later

dtzWill opened the pull request:
environment.sessionVariables cannot refer to the values of env vars,
and as a result this has caused problems in a variety of scenarios.

One use for these is that they're injected into /etc/profile,
elewhere these are used to populate an 'envfile' for pam
(`pam 5 pam_env.conf`) which mentions use of HOME being
potentially problematic.

Anyway if the goal is to make things easier for users,
simply do the NIX_PATH modification as extraInit.

This fixes the annoying problems generated by the current approach
(#40165 and others) while hopefully serving the original goal.

One way to check if things are borked is to try:

$ sudo env | grep NIX_PATH

Which (before this change) prints NIX_PATH variable with
an unexpanded $HOME in the value.

-------

This does mean the following won't contain user channels for 'will':
$ sudo -u will nix-instantiate --eval -E builtins.nixPath

However AFAICT currently they won't be present either,
due to unescaped $HOME.  Unsure if similar situation for other users
of sessionVariables (not sudo) work with current situation
(if they exist they will regress after this change AFAIK).



Tired of seeing warning, been running this change locally for a week
and seems to work for my machines.  Comments/testing appreciated.



<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

dtzWill committed:
NIX_PATH: don't prepend $HOME-based value in session variable, set later

environment.sessionVariables cannot refer to the values of env vars,
and as a result this has caused problems in a variety of scenarios.

One use for these is that they're injected into /etc/profile,
elewhere these are used to populate an 'envfile' for pam
(`pam 5 pam_env.conf`) which mentions use of HOME being
potentially problematic.

Anyway if the goal is to make things easier for users,
simply do the NIX_PATH modification as extraInit.

This fixes the annoying problems generated by the current approach
(#40165 and others) while hopefully serving the original goal.

One way to check if things are borked is to try:

$ sudo env | grep NIX_PATH

Which (before this change) prints NIX_PATH variable with
an unexpanded $HOME in the value.

-------

This does mean the following won't contain user channels for 'will':
$ sudo -u will nix-instantiate --eval -E builtins.nixPath

However AFAICT currently they won't be present either,
due to unescaped $HOME.  Unsure if similar situation for other users
of sessionVariables (not sudo) work with current situation
(if they exist they will regress after this change AFAIK).

RumataEstor commented on nixos/modules/services/misc/nix-daemon.nix:
```diff
@@ -446,6 +445,8 @@ in
             export NIX_REMOTE=daemon
         fi
+      '' + ''
+        export NIX_PATH="$HOME/.nix-defexpr/channels''${NIX_PATH:+:$NIX_PATH}"
```
Are these two single quotes necessary here?

dtzWill replied:
Within the double-single-quote they escape the dollar-sign (` '' ... ``${ .. ''`), if you're asking about the two single quotes in the middle of the line? Or are you suggesting dropping the outer `''` in favor of something else? I'm a fan of using `''` in most places for consistency if nothing else, but if there's a reason that's bad or there's a better alternative please LMK! :)

LnL7 replied:
I think adding a `test -e` here would also fix #40165.

We should probably either get this into 18.09 or revert https://github.com/NixOS/nixpkgs/pull/38351, this is going to cause a bunch of confusion otherwise.

dtzWill replied:
Agreed.  Not sure what you mean about 'test -e' but happy if you want to just add a commit doing so? :D

LnL7 replied:
Well nix shows a warning if NIX_PATH entries don't exist, so only extend it if ~/.nix-defexpr exists.

```bash
if [ -e "$HOME/.nix-defexpr/channels" ]; then
    export NIX_PATH="$HOME/.nix-defexpr/channels''${NIX_PATH:+:$NIX_PATH}"
fi
```

dtzWill replied:
Oh, right.  Thanks, updated!

grahamc replied:
Unfortunately this means the user needs to log out before the NIX_PATH gets updated after they add their first channel, but this is an improvement.

dtzWill commented:
Ping! Can we gather some examples/motivation for the current state of things? Ideally we can find a way to fix this breakage while still addressing those concerns...?

samueldr commented:
See today's discussions:

 * https://logs.nix.samueldr.com/nixos/2018-09-04#1536090584-1536091076;
 * https://logs.nix.samueldr.com/nixos-dev/2018-09-04#1536090981-1536091213;

dtzWill commented:
Should this be merged or are we waiting for something? Anyone have suggestions for how to proceed? Tests to run, people to check with?

Thanks!

dtzWill committed:
nix-daemon: only add channels dir to NIX_PATH if exists

Per reviewer comment (thanks!).

LnL7 reviewed (approved):
Looks good to me, I'd prefer this over reverting user channels back to nix-env only like before.

grahamc commented:
Pushed to master in 243e28bc96db0827eebfe2769de7e4393f791ca0, f3a114e088658786cfd5de5b2aa3e7cba9e96c64

It seems half this PR has already been applied to 18.09? I applied the second commit in 4c3a0ae44566b74ba4114e991f0e2ee0552caa87.
